### PR TITLE
feat(eap-items): read and filter array attributes natively from typed columns

### PR DESCRIPTION
**Pull request by phacops**
## Summary

Reading array attributes — both **selecting** and **filtering** — now goes natively through the typed `attributes_array_{string,int,float,bool}` map columns past the cutoff (`use_array_map_columns`) instead of the legacy `attributes_array` JSON column, with no `toJSONString` round-trip and no string coercion of stored elements. Returning/matching typed arrays directly is the whole point of double-writing arrays into typed `Map(String, Array(T))` columns.

Pre-cutoff data has no typed columns populated, so both paths fall back to the JSON column exactly as before. The data is double-written, so the two paths return the same results.

Linear: [EAP-562](https://linear.app/getsentry/issue/EAP-562/fix-array-storage-using-typed-columns-instead-of-json)

## Selecting arrays

We don't know an array attribute's element type at query time, and ClickHouse has no single column type that holds string/int/float/bool together, so a requested array attribute is selected as **four native `arrayElement(attributes_array_<type>, key)` sub-columns** (one per typed map column), using the non-bucketed access pattern (`arrayElement`, like `attributes_int`/`attributes_bool`). The four sub-columns are merged back into a single typed `val_array` in application code. Arrays are homogeneous, so exactly one sub-column is non-empty; element types (`val_str`/`val_int`/`val_double`/`val_bool`) are preserved natively — no `tuple`, no JSON. An absent array attribute (all four sub-columns empty) is returned as `is_null`, matching the legacy scalar behaviour.

Both read paths are retained:
- **TraceItemTable** / **GetTrace (per-attribute)**: the requested array attribute, read by key (typed sub-columns past the cutoff, else the JSON column).
- **TraceItemDetails** / **GetTrace (bulk)** / **ExportTraceItems**: every array attribute — the typed map columns whole past the cutoff, else the legacy `attributes_array` JSON-column allowlist.

## Filtering arrays

Array-membership (`<field>[*]:value` and its negation, plus `LIKE`/`NOT LIKE`, across string/number/boolean) keeps working — getsentry/sentry uses and tests it.

Sentry's EAP query builder sends **every** array-membership value as `val_str`, and the filter key only carries `TYPE_ARRAY` (never the element type), so the typed column to read can't be picked from the value's proto type. Instead, on the typed path the `val_str` is coerced to each native type it parses as and matched **natively** against that column, OR'd together:
- a numeric string (e.g. `"12"`) searches the int **and** float columns (it could be stored in either),
- `true`/`false` searches the bool column,
- every string searches the string column.

Matching uses each column's native element type — no stringifying the stored arrays (so e.g. `stack.colno[*]:12` matches the int `12` stored in `attributes_array_int`). `LIKE` patterns can only match string elements, so they read just `attributes_array_string`. Existence checks are `notEmpty(...)` over all four columns. Pre-cutoff, filtering uses the legacy `attributes_array` JSON column unchanged.

## Aggregations / group-by / order-by on arrays

- **`uniq` aggregation** ✅ kept (unchanged from master) — `FUNCTION_UNIQ` over an array attribute (e.g. crash-free user rate over `user_ids`) still resolves to `uniqArrayIfOrNull` over the legacy `attributes_array` JSON path. This is the one supported array aggregation and is independent of the typed-column read path.
- **Other aggregations** (sum/avg/count/…) ❌ rejected with a clear `BadSnubaRPCRequestException` (`_array_aggregation_to_expression`), same as master.
- **group-by / order-by** ❌ rejected in TraceItemTable request validation; the group-by/order-by builder no longer special-cases arrays. (These were non-functional/unused and would pull in keyset-pagination / SELECT-vs-GROUP-BY edge cases with array sub-columns.)

## Tests

- **Select** before/after cutoff: `test_endpoint_trace_item_table.py::test_select_array_column_before_and_after_cutoff`, `test_endpoint_trace_item_details.py::test_array_attributes_before_and_after_cutoff` + `test_convert_results_reads_typed_array_maps`, `test_endpoint_get_trace.py::test_process_results_merges_typed_array_subcolumns`.
- **Filter** native per-column matching: `test_common.py::TestTraceItemFiltersArrayMapColumns` (string-only for plain strings, int+float for numeric strings, bool for `true`/`false`, native rhs literals, no `arrayConcat`/`toString` over stored elements) + legacy JSON-path tests retained; ClickHouse-backed `test_endpoint_trace_item_table.py` array wildcard-search / equals tests (string ignore-case, int, all scalar rhs types).
- **Aggregations**: `test_aggregation.py::test_aggregation_to_expression_uniq_type_array` (uniq kept) + `..._sum_type_array_raises` (others rejected); `test_endpoint_time_series.py::test_crash_free_user_rate_with_array_attributes` (pre-existing, exercises uniq over arrays).
- **Rejections**: `test_endpoint_trace_item_table.py::TestArrayOperationsRejected` (group-by / order-by / non-uniq aggregation).
- `ruff` + affected unit tests pass locally; ClickHouse-backed integration tests run in CI.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_018AH9MHDdP2k6S52KpQvLo6

### CODE REVIEW

File: snuba/web/rpc/v1/resolvers/common/trace_item_table.py [hunk 1]
```diff
@@ -81,6 +100,10 @@ def _get_converter_for_type(
     elif key_type == AttributeKey.TYPE_DOUBLE:
         return lambda x: AttributeValue(val_double=float(x))
     elif key_type == AttributeKey.TYPE_ARRAY:
+        # Past the cutoff the SELECT returns a native typed-column tuple; otherwise the
+        # legacy JSON string.
```
[the remark this answers is not in the document]
Comment (phacops) [reply snuba/web/rpc/v1/resolvers/common/trace_item_table.py]: This is a false positive for this change. The aggregation array path was not switched to the typed columns: `aggregation_to_expression` resolves a `TYPE_ARRAY` field via `type_array_to_stored_array_json_path` (the legacy `attributes_array` JSON column) unconditionally — it does not produce the typed-column `tuple`. So for `FUNCTION_ANY` on an array the SELECT expression is still JSON-based, and the JSON converter selected here is the correct match. Passing `read_arrays_from_typed_columns` here would actually introduce the mismatch (tuple converter over a JSON expression). Only the `key`-column SELECT path (`_column_to_expression`) reads the typed tuple, and its converter is gated to match.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: snuba/web/rpc/v1/endpoint_export_trace_items.py [hunk 2]
```diff
@@ -290,7 +293,13 @@ def _build_query(
         ),
         SelectedExpression("attributes_int", column("attributes_int", alias="attributes_int")),
         SelectedExpression("attributes_bool", column("attributes_bool", alias="attributes_bool")),
-        *attributes_array_selected_expressions(),
+        # Past the cutoff, read every array attribute from the typed array map columns
+        # instead of the legacy attributes_array JSON-column allowlist.
+        *(
+            typed_array_map_selected_expressions()
+            if use_array_map_columns(in_msg.meta)
+            else attributes_array_selected_expressions()
```
[the remark this answers is not in the document]
Comment (phacops) [reply snuba/web/rpc/v1/endpoint_export_trace_items.py:309]: Good catch — real bug. Fixed: the cutoff is now evaluated against the routing-adjusted query window (`_export_query_meta(in_msg, routing_decision)`) in both `_build_query` (SELECT) and `_convert_rows` (decode), so the typed-column choice always matches the time window actually queried in the WHERE clause. (get_trace and trace-item details use a single meta throughout, so they aren't affected.)

---
_Generated by [Claude Code](https://claude.ai/code)_

File: snuba/web/rpc/v1/resolvers/R_eap_items/resolver_trace_item_table.py [hunk 3]
```diff
@@ -630,7 +682,11 @@ def build_query(
     if page_token_filter:
         additional_conditions.append(page_token_filter)
 
-    groupby = [_groupby_order_by_expression(attr_key) for attr_key in request.group_by]
+    groupby = [
+        expression
+        for attr_key in request.group_by
+        for expression in _groupby_order_by_expressions(attr_key, request.meta)
+    ]
```
[the remark this answers is not in the document]
Comment (phacops) [reply snuba/web/rpc/v1/resolvers/R_eap_items/resolver_trace_item_table.py]: False positive for TraceItemTable. Unlike export/details/get_trace, the table has **no allowlist** — each requested `TYPE_ARRAY` column is read by its specific key on both paths (typed sub-columns when `use_array_map_columns`, else `toJSONString(attributes_array.<key>)`), so no array attribute is ever dropped regardless of which path runs. The "non-allowlisted attributes are omitted" harm can't occur here.

It's also not a correctness bug:
- The SELECT and the decode (`convert_results`) gate on the **same** `request.meta`, so they never diverge — the sub-column merge runs iff the SELECT emitted sub-columns.
- Reading the legacy JSON column for an executed slice that's entirely post-cutoff is still correct: the typed columns are a *double-write*, so `attributes_array` is populated for all data. Gating on `request.meta` (the wider window) is conservative — it can pick the JSON read for a window that *starts* before the cutoff, but that returns identical data, never wrong/missing data.

So the only effect is occasionally choosing the JSON read instead of the (possibly faster) typed read for a routed slice. Aligning the table to the routed window would require re-gating *every* `use_array_map_columns(request.meta)` call in the resolver — including the load-bearing #8101 filter-membership path — on the routed window, for no correctness benefit. (Export gates on the routed window precisely because it *does* use the allowlist; the table has no equivalent gap.)

---
_Generated by [Claude Code](https://claude.ai/code)_

File: snuba/web/rpc/v1/resolvers/R_eap_items/resolver_trace_item_table.py [hunk 4]
```diff
@@ -270,22 +275,23 @@ def aggregation_filter_to_expression(
...
+    SELECT's `CAST(timestamp, 'Float64')`) so ClickHouse can read in primary-key order
+    while keeping the cast valid as a function of the grouped column.
     """
     if attr_key.name == "sentry.timestamp":
-        return snuba_column("timestamp")
-    return attribute_key_to_expression(attr_key)
+        return [snuba_column("timestamp")]
+    if attr_key.type == AttributeKey.Type.TYPE_ARRAY and use_array_map_columns(request_meta):
+        return list(type_array_typed_columns_select_expressions(attr_key))
+    return [attribute_key_to_expression(attr_key)]
```
[the remark this answers is not in the document]
Comment (phacops) [reply snuba/web/rpc/v1/resolvers/R_eap_items/resolver_trace_item_table.py]: False positive — this conflates the `SelectedExpression.name` with the SQL alias.

`_array_subcolumn_selected_expressions` sets `name=<label>.<col>` but the **expression** it wraps comes from `type_array_typed_columns_select_expressions`, whose alias is `<label_mapping_key>.<col>`. Only the expression alias is rendered as the SQL `AS`; the `name` is used solely by `transform_column_names` to rename the *result* column after execution (same as every non-array column: `name=column.label`, alias=`_build_label_mapping_key`). GROUP BY/ORDER BY reuse the same `type_array_typed_columns_select_expressions`, so they emit the identical alias.

Verified the built query — SELECT and GROUP BY/ORDER BY aliases are byte-identical:
```
SELECT  ...  AS `tags_TYPE_ARRAY.attributes_array_string`, ...
GROUP BY `tags_TYPE_ARRAY.attributes_array_string`, ...
ORDER BY `tags_TYPE_ARRAY.attributes_array_string`, ...
```
So ClickHouse matches them; no "not in GROUP BY". Applying the suggested change (basing the `name` on `_build_label_mapping_key`) would instead mislabel the *result* column (`tags_TYPE_ARRAY.attributes_array_string` instead of `tags.attributes_array_string`), breaking the converter's merge.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: snuba/web/rpc/v1/resolvers/R_eap_items/resolver_trace_item_table.py [hunk 5]
```diff
@@ -334,16 +340,15 @@ def _convert_order_by(
...
-                    direction=direction,
-                    expression=expression,
+            # covers `sentry.timestamp` ordering anywhere else.) GROUP BY uses the same
+            # expression(s) so the query stays valid.
+            for expression in _groupby_order_by_expressions(x.column.key, request_meta):
+                res.append(
+                    OrderBy(
+                        direction=direction,
+                        expression=expression,
+                    )
```
[the remark this answers is not in the document]
Comment (phacops) [reply snuba/web/rpc/v1/resolvers/R_eap_items/resolver_trace_item_table.py]: Ordering a TraceItemTable query by a `TYPE_ARRAY` attribute and paginating is a pre-existing unsupported combination, not a regression from this change.

The flex-time continuation (`FlexibleTimeWindowPageWithFilters.create` → `get_filters`) takes the last row's value for the order-by column and emits `column(alias) < literal(value)`. For an array column that `value` is the `val_array` itself, and `get_filters` does `literal(getattr(value, "val_array"))` — a protobuf `Array`, never a valid scalar continuation bound. So array sort + pagination already produced a broken/erroring continuation before this PR (back when the array was a single `tags_TYPE_ARRAY` column too); arrays were never a usable sort key for keyset pagination.

The 4-subcolumn expansion changes *which* error you'd hit on page 2, not whether that path ever worked. Making array columns genuinely paginable would need dedicated keyset handling (a per-typed-subcolumn tuple bound, decomposing the `val_array` back into typed values) in the pagination layer — out of scope here and not currently a supported operation. Non-paginated `ORDER BY <array>` still produces valid SQL (the four sub-column expressions match the SELECT). Happy to file a follow-up if array keyset pagination becomes a real requirement.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: snuba/web/rpc/common/common.py [hunk 6]
```diff
@@ -178,6 +180,60 @@ def attributes_array_selected_expressions() -> list[SelectedExpression]:
...
+    column order."""
+    merged: list[tuple[str, list[Any]]] = []
+    for base in bases:
+        elements: list[Any] = []
+        for typed_col in TYPED_ARRAY_MAP_COLUMNS:
+            values = row.pop(typed_array_select_subcolumn_name(base, typed_col), None)
+            if values:
+                elements.extend(values)
+        merged.append((base, elements))
+    return merged
```
Comment (onewland) [inline snuba/web/rpc/common/common.py:235]: have we considered (and rejected) not doing this for our upstream clients? 

I understand right now that we allow them to treat arrays as heterogeneous, and we need to do this mapping for a transition, but it would be nice if we could drop all this pre/post-processing later

**Review by onewland — APPROVED** [review]
I wasn't around for the discussion and implementation of separating types out in arrays, but this change seems fine and safe/quick to revert if we find a problem.

Will we be able to tell quickly if the change is broken?

File: snuba/web/rpc/common/common.py (hunk 6 above)
Comment (phacops) [reply snuba/web/rpc/common/common.py:235]: It's accepted we're not going to support heterogeneous arrays. The problem I'm trying to dance around here is we don't have a good type system for array attributes.

We defined a `TYPE_ARRAY` that doesn't tell you what scalar values you store, so when you query the array attribute, it has to look into the 4 columns and "merge", basically picking whatever value exists. And then this would pick the one that returned something and extend the results.

It's an evolution of a function that was trying to merge arrays with different types together so that's why it's complicated and I can definitely simplifies.

**Comment by phacops** [comment]
> Will we be able to tell quickly if the change is broken?

Yes, it would be fast, there are a few queries we can check after deploy in `s4s2` since SDKs have been sending arrays already.
